core/src/main/scala/caliban/execution/Executor.scala

```diff
@@ -37,26 +37,29 @@ object Executor {
   )(implicit trace: Trace): URIO[R, GraphQLResponse[CalibanError]] = {
     val wrapPureValues    = fieldWrappers.exists(_.wrapPureValues)
     val isDeferredEnabled = featureSet(Feature.Defer)
+    val isMutation        = request.operationType == OperationType.Mutation
 
     type ExecutionQuery[+A] = ZQuery[R, ExecutionError, A]
 
-    val execution = request.operationType match {
-      case OperationType.Query        => queryExecution
-      case OperationType.Mutation     => QueryExecution.Sequential
-      case OperationType.Subscription => QueryExecution.Sequential
+    val executionMode = queryExecution match {
+      case QueryExecution.Sequential => 0
+      case QueryExecution.Parallel   => 1
+      case QueryExecution.Batched    => 2
     }
 
     def collectAll[E, A, B, Coll[+V] <: Iterable[V]](
-      in: Coll[A]
+      in: Coll[A],
+      isTopLevelField: Boolean
     )(
       as: A => ZQuery[R, E, B]
     )(implicit bf: BuildFrom[Coll[A], B, Coll[B]]): ZQuery[R, E, Coll[B]] =
       if (in.sizeCompare(1) == 0) as(in.head).map(bf.newBuilder(in).+=(_).result())
+      else if (isTopLevelField && isMutation) ZQuery.foreach(in)(as)
       else
-        execution match {
-          case QueryExecution.Sequential => ZQuery.foreach(in)(as)
-          case QueryExecution.Parallel   => ZQuery.foreachPar(in)(as)
-          case QueryExecution.Batched    => ZQuery.foreachBatched(in)(as)
+        executionMode match {
```

Comment:
Does that really make a difference to use ints? Code would be cleaner without 😄

Reply:
Pattern matching on Ints in increments compiles to a tableswitch, which has O(1) performance regardless of the number of elements in the pattern matching. Plus I do think that equality on ints is less complex than on objects since there's no need for type-checking.

Having said that, we only have 3 elements, and this code is probably a millionth of the overall execution complexity so I'll just go back to making it more readable 😅

Reply:
Made the change. I placed Sequential last now though since I find it very hard to believe anyone would ever want to use that.. right?

Reply:
Thanks 😄